Add support for switching the scheme based on the app's theme

bennettnicholas opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This pull request solved the problem of users not being able to set color schemes specifically for dark or light mode. Now the code has been updated to accept a dark and light color scheme in the json. The old setting is still compatible. Keep in mind if you update your color scheme through the settings UI, it will set both dark and light to the color scheme selected. This is because the settings UI update for selecting both Dark and Light color schemes is not supported yet.

This also solves the problem of the UI not using the system OS theme. Now you can select system theme and your color scheme will be selected based on if the system theme is dark or light.


<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References
#4066 

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Closes #4066 
* [x] Closes #14050
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA. 
* [x] Tests added/passed I believe so, added one test to ColorSchemeTests.cpp and I believe it passed. Also had to modify TerminalSettingsTests.cpp to accept the new ApplyAppearanceSettings function template
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [x] Schema updated.
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #4066 and also teams messages with @carlos-zamora 

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
-Removed ColorSchemeName from MTSMSettings.h in order to process the setting for both string and object.
-Added DarkColorSchemeName and LightColorSchemeName properties to the AppearanceConfig to replace ColorSchemeName.
-Hacked a few processes to play nice with all 3 properties listed above as in some cases around the UI, we need to still use the ColorSchemeName. Once we change the UI I believe we can go back to just Dark and LightColorSchemeName
-Added and Updated Test to align to the new code.

Acceptable Json values,

"colorScheme": 
                {
                    "dark": "Campbell",
                    "light": "Campbell"
                }
or

"colorScheme": "Campbell"

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
Individual testing along with the test case added.

carlos-zamora commented:
Hi @bennettnicholas,

Chatted with the team today. Here's some notes from our discussion:
- don't worry about the color scheme not changing _when_ you update the OS theme. We can throw that in a follow-up.
   - FYI: #1125 would probably work. It's WM_SETTINGCHANGE should tell us when a setting changes and which one.
- Settings UI:
   - I'm ok with this being another case (like window padding) where we don't overwrite the setting if the user didn't change it. It's not great at all, but it should suffice until we get the new one designed (others may disagree here, so we'll leave that up for discussion when they review this PR). Just make sure we track this in #10000.

carlos-zamora commented:
@bennettnicholas FYI found this this (sometimes) breaks color schemes: #14174

Please take a look when you get the chance. 😊

bennettnicholas commented:
@carlos-zamora The bug has been identified and fixed. Along with another bug was discovered.  Thanks for catching the bug.

lhecker commented on src/cascadia/TerminalSettingsModel/AppearanceConfig.cpp:
```diff
@@ -79,6 +102,20 @@ void AppearanceConfig::LayerJson(const Json::Value& json)
     JsonUtils::GetValueForKey(json, OpacityKey, _Opacity, JsonUtils::OptionalConverter<double, IntAsFloatPercentConversionTrait>{});
+    if (json["colorScheme"].isString())
+    {
+        // to make the UI happy, set ColorSchemeName.
+        JsonUtils::GetValueForKey(json, ColorSchemeKey, _ColorSchemeName);
+        JsonUtils::GetValueForKey(json, ColorSchemeKey, _DarkColorSchemeName);
+        JsonUtils::GetValueForKey(json, ColorSchemeKey, _LightColorSchemeName);
```
```suggestion
        JsonUtils::GetValueForKey(json, ColorSchemeKey, _ColorSchemeName);
        _DarkColorSchemeName = _ColorSchemeName;
        _LightColorSchemeName = _ColorSchemeName;
```

lhecker commented on src/cascadia/TerminalSettingsModel/AppearanceConfig.cpp:
```diff
@@ -79,6 +102,20 @@ void AppearanceConfig::LayerJson(const Json::Value& json)
+    else if (json["colorScheme"].isObject())
+    {
+        // to make the UI happy, set ColorSchemeName to whatever the dark value is.
+        JsonUtils::GetValueForKey(json["colorScheme"], "dark", _ColorSchemeName);
+        JsonUtils::GetValueForKey(json["colorScheme"], "dark", _DarkColorSchemeName);
+        JsonUtils::GetValueForKey(json["colorScheme"], "light", _LightColorSchemeName);
```
```suggestion
        JsonUtils::GetValueForKey(json["colorScheme"], "dark", _DarkColorSchemeName);
        JsonUtils::GetValueForKey(json["colorScheme"], "light", _LightColorSchemeName);
        _ColorSchemeName = _DarkColorSchemeName;
```

lhecker reviewed (commented):
Looks pretty good to me! Just a few comments/questions...

bennettnicholas commented on src/cascadia/TerminalSettingsModel/AppearanceConfig.cpp:
```diff
@@ -50,6 +54,25 @@ Json::Value AppearanceConfig::ToJson() const
+            JsonUtils::SetValueForKey(json["colorScheme"], "light", _LightColorSchemeName);
+        }
+    }
+    else if (HasColorSchemeName())
+    {
+        JsonUtils::SetValueForKey(json["colorScheme"], "dark", _ColorSchemeName);
+        JsonUtils::SetValueForKey(json["colorScheme"], "light", _ColorSchemeName);
+    }
```
I dont believe we can shorten this, as it wouldnt be able to keep the same logic. We have to first check if the light and dark are set, if not then we need to move onto the next check.

bennettnicholas commented:
> Looks pretty good to me! Just a few comments/questions...

@lhecker sorry for the delay, been fairly busy. I was able to incorporate most of the comments. The biggest thing around some of the logic in the code is because we are hacking the backend together until we update the UI. Once the UI is updated, a lot of this logic can be dropped. In many cases we have to process the color scheme name differently depending on if it's coming from the JSON or the UI. Please take a look at some of the changes and let me know what you think.

bennettnicholas commented on src/cascadia/TerminalSettingsModel/AppearanceConfig.cpp:
```diff
@@ -50,6 +54,25 @@ Json::Value AppearanceConfig::ToJson() const
+    if (HasDarkColorSchemeName() && HasLightColorSchemeName())
+    {
+        // check if the setting is coming from the UI, if so grab the ColorSchemeName until the settings UI is fixed..
+        if (_ColorSchemeName != _DarkColorSchemeName)
```
Is this talking about line 57 or 60?

lhecker replied:
Line 60.

bennettnicholas replied:
In that case, it would change the logic. We are checking here if the ColorSchemeName differs. This is because ColorSchemeName is still being used by the UI. We check that value differs, indicating the change came from the UI. We have to then can choose to use the old variable for saving to the JSON.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/TerminalSettings.cpp:
```diff
@@ -183,13 +183,49 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
-        if (!appearance.ColorSchemeName().empty())
+
+        const auto requestedTheme = currentTheme.RequestedTheme();
+        const auto defaultName = appearance.ColorSchemeName();
```
Yeah, I don't think we need `defaultName` (`ColorSchemeName()`) anymore. I see that you're keeping it around so that the UI has a value it's in charge of, but I think the right move there should be that the TerminalSettingsEditor sets `DarkColorSchemeName` and `LightColorSchemeName` when it's set via the UI.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/CascadiaSettings.cpp:
```diff
@@ -326,6 +326,9 @@ Model::Profile CascadiaSettings::DuplicateProfile(const Model::Profile& source)
         DUPLICATE_SETTING_MACRO_SUB(appearance, target, Opacity);
+        DUPLICATE_SETTING_MACRO_SUB(appearance, target, DarkColorSchemeName);
+        DUPLICATE_SETTING_MACRO_SUB(appearance, target, LightColorSchemeName);
+        DUPLICATE_SETTING_MACRO_SUB(appearance, target, ColorSchemeName);
```
`ColorSchemeName` was already added in the macro above because it's a part of `MTSM_APPEARANCE_SETTINGS` in MTSMSettings.h.

Since that's the case, maybe we should be replacing it with `DarkColorSchemeName` and `LightColorSchemeName` there too?

carlos-zamora commented on src/cascadia/inc/ControlProperties.h:
```diff
@@ -16,10 +16,11 @@
 #define CONTROL_APPEARANCE_SETTINGS(X)                                                                                                          \
     X(til::color, SelectionBackground, DEFAULT_FOREGROUND)                                                                                      \
     X(double, Opacity, 1.0)                                                                                                                     \
+    X(winrt::hstring, ColorSchemeName, L"Campbell")                                                                                             \
```
I don't think we need this at all. `ICoreAppearance` should be able to handle all of the values for the color table. So in this layer, we shouldn't even have/need a concept of a "color scheme name" at all

carlos-zamora reviewed (changes requested):
- [x] please update the schema
- [ ] please update the docs
- [x] We shouldn't need `ColorSchemeName` in the settings model anymore. We should only need the light/dark variants. Upon loading the JSON, if we encounter `"colorScheme": "<name>"`, we should copy that name into both variants. We'll probably need special handling in `Profile.cpp::LayerJson` to accomplish this.
- [x] I think you need to modify `MTSMSettings.h` more. This is where we do the magic of defining a new setting "once" and it goes basically everywhere in the settings model (removes the need to implement a lot of boiler plate code)
- [ ] update `TerminalSettingsEditor`. Specifically, the UI shouldn't change, but the text passed in should directly set the dark/light variant. Urgh, this still puts us in the situation of us not being sure which one to display depending on the theme. I say, look at the theme for the settings model that is in use, and display the relevant color scheme name. It's a bit weird, but I think that's a good intermediate step. CC @DHowett

zadjii-msft commented:
Alrighty so we did a quick little bugs bash with the team this afternoon, and pulled this in. Two notes:
* [ ] _Something we'll follow up on, but doesn't deserve to block_: If a profile has the same scheme for light and dark, then we should just re-serialize as a single string, and not expand out to the whole object. That will leave most profiles untouched, so users should only see the expanded object if they actually opt-in to using this feature. 
* [ ] _Something we need to discuss more as a team_: As currently implemented, this strictly obeys `theme.window.applicationTheme`. If the user's theme is set to `light` or `dark`, then this feature will _never_ work, because those inbox themes always set `window.applicationTheme`. Is that what we want? Probably not? Maybe this should _always_ obey the OS theme setting, regardless of the app theme setting. (seperately, there's _a bullet point in #3327_ about changing the **theme** based off the OS theme. That seems relevant to the discussion. However, most folks would still only have `theme: dark` as their theme, so they'd need to manually change that setting too.... something we need to discuss more.)


However, I do think this is really cool and I'm excited to see this merge. These might be points where we decide to ultimately just ingest now, and fix in post.

bennettnicholas commented:
Hey Guys, sorry for the delay. I just got back from PTO. Ill look into this soon, I am still trying to catch back up on life. Hopefully later this week I can start addressing these comments.

carlos-zamora commented:
Linking a relevant thread: https://github.com/microsoft/terminal/issues/4066#issuecomment-1311947499

Still keeping needs-discussion tag though. More just to track if/how to get that functionality right because `"theme": "dark"` is now the default, which means it's blocking `"theme": "system"` unless explicitly set by the user.

zadjii-msft commented:
IMO, I think that's fine. Put it clearly in the documentation for the feature. Make a new docs page: "Tutorial: Sync the terminal colors to the OS theme". Link that to the theme docs, and the scheme docs. Blamo. Because there was one person in that thread that:

> Then, however I change the application theme (directly in Terminal settings, or by changing my system theme) I get a scheme that works for me. I can even set both schemes to be the same if I don't want them to change depending on app theme.

which also sounds like "I want to be able to change the Terminal theme and have the scheme update to match". That all seems possible. Two knobs, but, okay by me (and seemingly the folks in that thread)

zadjii-msft commented:
![live-scheme-switching-000](https://user-images.githubusercontent.com/18356694/202005782-1869f3a4-2d8f-48d2-86e6-3da9059aa6ef.gif)

I'm gonna start pushing some commits here to get this across the finish line

DHowett commented on src/cascadia/TerminalApp/AppLogic.idl:
```diff
@@ -60,6 +60,7 @@ namespace TerminalApp
         void Quit();
 
         void LoadSettings();
+        void ReloadSettings();
```
Having both of these moves the burden of knowing when to use which up to the caller. Can they be consolidated? Can `ReloadSettings` just implicitly do a `LoadSettings` if it's the first one?

zadjii-msft commented on src/cascadia/TerminalSettingsEditor/Appearances.xaml:
```diff
@@ -39,10 +39,11 @@
                        Style="{StaticResource TextBlockSubHeaderStyle}" />
 
             <!--  Color Scheme  -->
+            <!--  This currently only display the Dark color scheme, even if the user has a pair of schemes set.  -->
```
FWIW, I think we should just do the whole of SUI changes separately as a blocking codehealth follow-up. These changes are pretty big already.

DHowett replied:
Okay, it's on you to file that and drive it down. Thanks! :smile:

zadjii-msft commented on src/cascadia/TerminalSettingsEditor/Appearances.cpp:
```diff
@@ -370,7 +370,8 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
     void Appearances::CurrentColorScheme(const ColorScheme& val)
     {
-        Appearance().ColorSchemeName(val.Name());
+        Appearance().DarkColorSchemeName(val.Name());
```
Wouldn't that be handled by

```c++
                else if (settingName == L"DarkColorSchemeName" || settingName == L"LightColorSchemeName")
                {
                    _PropertyChangedHandlers(*this, PropertyChangedEventArgs{ L"CurrentColorScheme" });
                }
```

up on L246?

DHowett replied:
I don't believe so. You're going into the inner TSM Appearance object, which has no notification capabilities (?), and changing the datamodel field directly. Unless I am looking at the wrong layer, and Appearance here is an AppearanceViewModel? If so, disregard, but then know that this will trigger a spurious double-notification :)

zadjii-msft replied:
Are we sure? cause If I add 
```c++
_PropertyChangedHandlers(*this, PropertyChangedEventArgs{ L"CurrentColorScheme" });
```
that causes the app to crash when changing the scheme in the UI

carlos-zamora replied:
> I don't believe so. You're going into the inner TSM Appearance object, which has no notification capabilities (?), and changing the datamodel field directly. Unless I am looking at the wrong layer, and Appearance here is an AppearanceViewModel? If so, disregard, but then know that this will trigger a spurious double-notification :)

For completeness of this thread, Appearance was an AppearanceViewModel.

zadjii-msft commented:
Last commit closes #14050 too

zadjii commented:
Ahhh bot, it's the holidays

DHowett commented on src/cascadia/TerminalSettingsModel/defaults.json:
```diff
@@ -103,6 +103,28 @@
             "brightWhite": "#F2F2F2"
         },
+        {
+            "name": "Campbell Light",
```
we're adding a new stock color scheme? I thought we weren't gonna do that until we got scheme healing. Hmm...

zadjii-msft replied:
Fair point. I thought it made sense, but yea, default schemes annoying.

DHowett reviewed (approved):
Questions, but... I'm not gonna block it any longer.

@zadjii-msft before you merge this, would you please give the PR a title in the imperative form s.t. it could complete the sentence "when merged, this commit will ..."?

carlos-zamora commented on src/cascadia/TerminalSettingsModel/AppearanceConfig.cpp:
```diff
@@ -50,6 +54,19 @@ Json::Value AppearanceConfig::ToJson() const
     JsonUtils::SetValueForKey(json, SelectionBackgroundKey, _SelectionBackground);
     JsonUtils::SetValueForKey(json, CursorColorKey, _CursorColor);
     JsonUtils::SetValueForKey(json, OpacityKey, _Opacity, JsonUtils::OptionalConverter<double, IntAsFloatPercentConversionTrait>{});
+    if (HasDarkColorSchemeName() && HasLightColorSchemeName())
```
Hmm... If the JSON has...
```js
"colorScheme": {
   "dark": "foo"
// omit "light"
}
```

And you hit the save button in SUI, you lose the entire "colorScheme" object entirely.

I think we need this statement to be `||` and we need to check if each one exists before writing it.

Normally I wouldn't care, but this is data loss so I think it's kinda important.

carlos-zamora commented on src/cascadia/TerminalSettingsEditor/ColorSchemeViewModel.cpp:
```diff
@@ -52,7 +52,9 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
     bool ColorSchemeViewModel::IsDefaultScheme()
     {
-        return _Name == _settings.ProfileDefaults().DefaultAppearance().ColorSchemeName();
+        auto defaultAppearance = _settings.ProfileDefaults().DefaultAppearance();
```
```suggestion
        const auto defaultAppearance = _settings.ProfileDefaults().DefaultAppearance();
```

carlos-zamora reviewed (changes requested):
@PankajBhojwani FYI this may conflict with #14470 

Blocking over the JSON bug

carlos-zamora reviewed (approved):
Awesome. Thanks @zadjii-msft and @bennettnicholas!

carlos-zamora commented:
@zadjii-msft don't forget to document this plz

MatejKafka commented:
Is there a way to use a dark *theme*, but still switch the color *scheme* of a profile based on the system-wide theme?

zadjii-msft commented:
yes and no.

You could do something like

```
    "theme": { "dark": "my light", "light": "dark" },
    "themes":
    [
        {
            "name": "my light",
            "tab":
            {
                "background": "terminalBackground",
                "unfocusedBackground": "#00000000"
            },
            "tabRow":
            {
                "background": "#2e2e2e",
                "unfocusedBackground": "#333333FF"
            },
            "window":
            {
                "applicationTheme": "light"
            }
        },
```

so that the tabs were dark, even if the rest of the UI was _light_. The `applicationTheme` being `light` would them let the `scheme` pick the `light` value.



<hr>

But more completely - trying to support that scenario (dark UI, OS-following color scheme) proved a bit to confusing for the settings to support, hence the current approach. (as discussed around https://github.com/microsoft/terminal/issues/4066#issuecomment-1311901715)
